Add 'percentile' to template modes and plot_unit_templates

alejoe91 opened the pull request.

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1206,57 +1217,68 @@ def precompute_templates(self, modes=("average", "std")) -> None:
                 else:
                     raise ValueError("mode must in median/average/std")
-                self._template_cache[mode][unit_ind][:, mask] = arr
+                self._template_cache[mode_names[mode]][unit_ind][:, mask] = arr
```
In L1225, the ValueError should now include percentile as well.

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1206,57 +1217,68 @@ def precompute_templates(self, modes=("average", "std")) -> None:
+                    assert 0 <= percentile <= 100, "percentile must be between 0 and 100 inclusive"
+                    arr = np.percentile(wfs, percentile, axis=0)
                 else:
                     raise ValueError("mode must in median/average/std")
```
```suggestion
                    raise ValueError(f"'mode' must be in {_possible_template_modes}")
```

zm711 reviewed (commented):
Actually it's letting me suggest on some lines, but not others.

alejoe91 commented:
@samuelgarcia I also added a ScaleWidget to the `plot_waveforms/templates` and the shading also to the mpl/ipy backend.

Sample sortingview plot: https://figurl.org/f?v=gs://figurl/spikesortingview-11dev&d=sha1://fb1fc49ee6c9419f55b2ee9d2ce9daf29aa23ffd&zone=aind

Sample ipywidgets: 
[Screencast from 14-11-2023 15:54:17.webm](https://github.com/SpikeInterface/spikeinterface/assets/17097257/098e8149-99e8-473c-bc1b-38f3fc65f161)

Also added an automatic darker color for templates in case that waveforms are plotted

samuelgarcia commented on src/spikeinterface/widgets/unit_waveforms.py:
```diff
@@ -225,9 +241,71 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
+                        )
+                        upper_bound = (
+                            dp.templates_shading[3][i, :, :][:, chan_inds] * dp.y_scale + dp.y_offset[:, chan_inds]
+                        )
```
This could be a general case of N must be even. and written in a loop

alejoe91 replied:
done

samuelgarcia commented on src/spikeinterface/widgets/unit_waveforms.py:
```diff
@@ -274,8 +352,9 @@ def plot_ipywidgets(self, data_plot, **backend_kwargs):
-        self.unit_selector = UnitSelector(data_plot["unit_ids"])
+        self.unit_selector = UnitSelector(data_plot["unit_ids"], layout=widgets.Layout(height="80%"))
         self.unit_selector.value = list(data_plot["unit_ids"])[:1]
+        self.scaler = ScaleWidget(layout=widgets.Layout(height="20%"))
```
Maybe we could expose the initial value of the sclae (default 1) also for matpltolib

alejoe91 replied:
added it for matplotlib and ipywidgets. Moved the scaling mechanism to `plot_matplotlib`

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1171,12 +1171,20 @@ def get_waveforms_segment(self, segment_index: int, unit_id, sparsity):
-    def precompute_templates(self, modes=("average", "std")) -> None:
+    def precompute_templates(self, modes=("average", "std", "median", "percentile"), percentile=None) -> None:
         """
         Precompute all template for different "modes":
```
```suggestion
        Precompute all templates for different "modes":
```

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1171,12 +1171,20 @@ def get_waveforms_segment(self, segment_index: int, unit_id, sparsity):
+        percentile: float, default: None
+            Percentile to use for mode="percentile"
 
         The results is cache in memory as 3d ndarray (nunits, nsamples, nchans)
```
```suggestion
        The results is cached in memory as a 3d ndarray (nunits, nsamples, nchans)
```

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1171,12 +1171,20 @@ def get_waveforms_segment(self, segment_index: int, unit_id, sparsity):
+            Percentile to use for mode="percentile"
 
         The results is cache in memory as 3d ndarray (nunits, nsamples, nchans)
         and also saved as npy file in the folder to avoid recomputation each time.
```
```suggestion
        and also saved as an npy file in the folder to avoid recomputation each time.
```

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1206,57 +1217,68 @@ def precompute_templates(self, modes=("average", "std")) -> None:
-    def get_all_templates(self, unit_ids: Optional[Iterable] = None, mode="average"):
+    def get_all_templates(self, unit_ids: Optional[Iterable] = None, mode="average", percentile: float | None = None):
         """
         Return  templates (average waveform) for multiple units.
```
```suggestion
        Return templates (average waveforms) for multiple units.
```

zm711 replied:
Or 

```suggestion
        Return the template (average waveform) for multiple units.
```

zm711 commented on src/spikeinterface/core/waveform_extractor.py:
```diff
@@ -1206,57 +1217,68 @@ def precompute_templates(self, modes=("average", "std")) -> None:
         unit_ids: list or None
             Unit ids to retrieve waveforms for
-        mode: "average" | "median" | "std", default: "average"
+        mode: "average" | "median" | "std" | "perentile", default: "average"
```
```suggestion
        mode: "average" | "median" | "std" | "percentile", default: "average"
```

zm711 commented on src/spikeinterface/widgets/unit_waveforms.py:
```diff
@@ -48,6 +48,18 @@ class UnitWaveformsWidget(BaseWidget):
+        If None, the shading is +/- the standard deviation of the templates.
+        If float, it controls the percentile of the template values used to shade the templates.
+        Note that it is one-sided: if 5 is given, the 5th and 95th percentiles are used to shade
+        the templates. If list of floats, it can be a 2-element or 4-element list.
```
```suggestion
        the templates. If list of floats, it can be a 2-element or a 4-element list.
```

zm711 reviewed (commented):
A few more docstring cleanup comments
